message_processor.go

```diff
@@ -50,36 +83,48 @@ func consumeLog(lf *logPair) {
 			logrus.WithField("id", lf.info.ContainerID).WithError(err).Debug("Ignoring error")
 			dec = protoio.NewUint32DelimitedReader(lf.stream, binary.BigEndian, 1e6)
 		}
-		if sendMessage(lf.splunkl, &buf, lf.info.ContainerID) == false {
-			continue
-		}
-		if sendMessage(lf.jsonl, &buf, lf.info.ContainerID) == false {
-			continue
-		}
 
+		if shouldSendMessage(buf.Line) {
+			// Append to temp buffer
+			appendToTempBuffer(tmpBuf, &buf)
+			// Check for temp buffer timer expiration
+			diff := time.Now().Sub(tmpBuf.bufferTimer)
+			if diff > tempMsgBufferHoldDuration {
+				tmpBuf.bufferHoldDurationExpired = true
+			}
+
+			sendMessage(lf.splunkl, &buf, tmpBuf, lf.info.ContainerID)
+			sendMessage(lf.jsonl, &buf, tmpBuf, lf.info.ContainerID)
+			//temp buffer and values reset
+			if tmpBuf.bufferReset {
+				resetTempBufferValues(tmpBuf)
+			}
+		}
 		buf.Reset()
 	}
-
 }
 
 // send the log entry message to logger
-func sendMessage(l logger.Logger, buf *logdriver.LogEntry, containerid string) bool {
+func sendMessage(l logger.Logger, buf *logdriver.LogEntry, tBuffer *tmpBuffer, containerid string) {
 	var msg logger.Message
-	if !shouldSendMessage(buf.Line) {
-		return false
-	}
+	if !buf.Partial || tBuffer.bufferHoldDurationExpired || tempMsgBufferMaximum <= tBuffer.tBuf.Len() {
```

Comment:
If `buf.Partial` is false, do we do an immediate `l.Log` ? is `l.Log` where we do the event batching ?

Reply:
refactor out a member function of the buf length check
```
func (b *tmpBuffer) MaxLengthReached() bool {
    return tempMsgBufferMaximum <= tBuffer.tBuf.Len()
}
```

Reply:
Refactor out another member function say `ShouldFlush` for `tmpBuffer`
```
func (b *tmpBuffer) ShouldFlush() bool {
    return b.TimedOut || b.MaxLengthReached()
}
```

Here the condition check becomes
```
if buf.Partial || tBuffer.ShouldFlush() {
...
}
```

Reply:
If buf.Partial is false we collect the message in the temp buffer. So messages are appended in the temp buffer until one of the (3) conditions are triggered.
The batching of messages itself is done later. The values that we are setting at this moment (like temp buffer hold duration of 100 ms) can change later based on our performance tests.